### PR TITLE
Fix for incorrect line drawing beyond 32768 row or column

### LINKED ISSUE #4555: cv::line draws wrong lines when coordinates are very far from image
Transferred from http://code.opencv.org/issues/2988

```
|| Adi Shavit on 2013-04-25 20:48
|| Priority: Normal
|| Affected: 2.4.0 - 2.4.4
|| Category: core
|| Tracker: Bug
|| Difficulty: None
|| PR: 
|| Platform: None / None
```
## cv::line draws wrong lines when coordinates are very far from image 

```
There seems to be a bug with cv::line related to clipping. 
Sometimes, when it gets points outside and very far from the image it draws completely wrong lines.

Here's some code to demonstrate this:

   {
      cv::Mat img = cv::Mat::zeros(cv::Size(480,360), CV_8UC3);   

      // The points
      Point p2(262, 23),  q2(91618510, -54756195);
      Point p3(263, 214),  q3(80120779, 11846827);

      // The bad lines
      line(img, p2, q2, cv::Scalar(255,255,255));
      line(img, p3, q3, cv::Scalar(255,255,255));

      // Clip BEFORE drawing
      cv::clipLine(img.size(), p2, q2);
      cv::clipLine(img.size(), p3, q3);

      // Good lines
      line(img, p2, q2, cv::Scalar(0,255,255));
      line(img, p3, q3, cv::Scalar(0,255,255));

      imshow("img", img);
      waitKey();

   }

This also happens with CV_AA.
```
##### Gurpinder Sandhu on 2013-04-26 14:53

```
@Vadim I think this can solved simply by adding an assertion in the line function:

@
CV_Assert(pt1.x < img.cols && pt1.y < img.rows);
CV_Assert(pt2.x < img.cols && pt2.y < img.rows);
@

where pt1, pt2 are the input points and img is the input image.
Please give your comments
```
##### Adi Shavit on 2013-04-30 17:59

```
@ Gurpinder Sandhu, I don't think it should assert that points are inside the image. For several reasons:
# First, it already works with internal clipping for many cases.
# Sometimes, you need to draw lines to points outside the image. Auto clipping is very helpful for making clean code.
# Since the Bresenham algorithm discretizes the line. A line to the clipped point will generally not be the same as the line to the point outside the image.
```
##### Anna Kogan on 2013-05-07 08:16

```
Hello Gurpinder,
Thank you for reporting the issue. I agree with Adi that asserts are not needed in the case. If you could find another solution on your side, a "pull request":http://www.code.opencv.org/projects/opencv/wiki/How_to_contribute in our GitHub repo would be highly appreciated!
```

### LINKED ISSUE #5473: Can't draw to image to coordinates >32768
I've found a strange issue while working with big images. I noticed things don't get drawn to coordinates above 32768, even though the the drawing function expects a cv::Point which should have int type (and thus, minimum 32 bit on my configuration) and so it should support values much higher.

Example code:

``` .cpp
cv::Mat outImage = cv::Mat(5, 35000, CV_8UC3); // a really wide image
cv::rectangle(outImage, cv::Point(32500, 0), cv::Point(32600, 3), cv::Scalar(200, 0, 0), CV_FILLED, 8); // This gets drawn
cv::rectangle(outImage, cv::Point(33000, 0), cv::Point(33100, 3), cv::Scalar(200, 0, 0), CV_FILLED, 8); // This doesn't get drawn
```

I found the issue using the cv::rectangle() function but I suspect it concerns several other drawing functions, like cv::line() according to this:
http://answers.opencv.org/question/9284/drawing-limited-by-size-of-int-in-opencv/

They mention a workaround to use image ROIs when working with such coordinates, which works, but still, I didn't see this issue coming so I assumed the bug was in my code and spent some time debugging, so there should be at least some warning about this in the documentation.

Using OpenCV 3.0, 64 bit, "stock" binaries.

### PR DESCRIPTION
**Pull request by terfendail**
resolves #5473
resolves #4555
### This pullrequest changes

Fix line drawing implementation to handle coordinates above 32767 correctly

### CODE REVIEW

**Comment by vpisarev** [comment]
@terfendail, thanks for the patch! Even though we cannot say that the drawing functions are performance-critical, I would like to keep int's if possible, not to change everything to int64/double.  I believe, some people may want to use on Raspberry Pi, Quarks etc. where double precision is slow or unavailable. One possible way to expand the value range and yet keep it within 32-bit integers is to slightly reduce the number of fractional bits; perhaps, 16 fractional bits in fixed-point representation of coordinates is too much and maybe we can reduce it to 15 or even 14 bits. Could you explore such solution?

**Comment by terfendail** [comment]
Unfortunately reduction of fractional bit number cause failure of related tests(there are a few points for diagonal lines that are drawn shifted by one pixel).
Since there is a workaround to use matrix ROI to draw beyond 32768 it could be built into all drawing functions. So if the shape goes out of 32768 boundary it will be drawn on subrect of the image. The solution doesn't allow to draw very big shapes but at least allow to draw small shapes anywhere without special treatment from user side.

**Comment by vrabaud** [comment]
there should at least be an assert right?

File: modules/imgproc/src/drawing.cpp [hunk 1]
```diff
@@ -79,16 +79,29 @@ FillConvexPoly( Mat& img, const Point* v, int npts,
...
-    if( img_size.width <= 0 || img_size.height <= 0 )
+    if (img_size.width <= 0 || img_size.height <= 0)
         return false;
 
-    x1 = pt1.x; y1 = pt1.y; x2 = pt2.x; y2 = pt2.y;
+    int64 &x1 = pt1.x, &y1 = pt1.y, &x2 = pt2.x, &y2 = pt2.y;
+    int c1, c2;
+    img_size.width--;
+    img_size.height--;
+    int64 &right = img_size.width, &bottom = img_size.height;
```
Comment (alalek) [inline modules/imgproc/src/drawing.cpp]: Why are separate decrement statements and references used?

Comment (terfendail) [inline modules/imgproc/src/drawing.cpp]: fixed

**Comment by vpisarev** [comment]
great! let's finally merge it in :+1:

**Comment by terfendail** [comment]
Backported as PR #8031

**Pull request merged**